pkg/serializer/internal/metrics/iterable_series.go

```diff
@@ -110,23 +110,153 @@ func describeItem(serie *metrics.Serie) string {
 // If a compressed payload is larger than the max, a new payload will be generated. This method returns a slice of
 // compressed protobuf marshaled MetricPayload objects.
 func (series *IterableSeries) MarshalSplitCompress(bufferContext *marshaler.BufferContext, config config.Component, strategy compression.Component) (transaction.BytesPayloads, error) {
-	var err error
-	var compressor *stream.Compressor
+	pb, err := series.NewPayloadsBuilder(bufferContext, config, strategy)
+	if err != nil {
+		return nil, err
+	}
+
+	err = pb.startPayload()
+	if err != nil {
+		return nil, err
+	}
+
+	// Use series.source.MoveNext() instead of series.MoveNext() because this function supports
+	// the serie.NoIndex field.
+	for series.source.MoveNext() {
+		err = pb.writeSerie(series.source.Current())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// if the last payload has any data, flush it
+	err = pb.finishPayload()
+	if err != nil {
+		return nil, err
+	}
+
+	return pb.payloads, nil
+}
+
+// MarshalSplitCompressMultiple uses the stream compressor to marshal and compress one series into two sets of payloads.
+// One set of payloads contains all metrics, and the other contains only those that pass the provided filter function.
+// This function exists because we need a way to build both payloads in a single pass over the input data, which cannot be iterated over twice.
+func (series *IterableSeries) MarshalSplitCompressMultiple(config config.Component, strategy compression.Component, filterFunc func(s *metrics.Serie) bool) (transaction.BytesPayloads, transaction.BytesPayloads, error) {
+	bufferContext := marshaler.NewBufferContext()
```

Comment:
Why do we create the buffer context here on behalf of the caller, but require them to pass it when calling `MarshalSplitCompress`?